discretisedfield/util/util.py

```diff
@@ -17,71 +18,64 @@ def array2tuple(array):
     return array.item() if array.size == 1 else tuple(array.tolist())
 
 
-def as_array(mesh, dim, val):
-    array = np.empty((*mesh.n, dim))
-    if isinstance(val, numbers.Real) and (dim == 1 or val == 0):
-        # The array for a scalar field with numbers.Real value or any
-        # field with zero value.
-        array.fill(val)
-    elif isinstance(val, complex) and (dim == 1 or val == 0):
-        array = np.empty((*mesh.n, dim), dtype='complex128')
-        array.fill(val)
-    elif isinstance(val, np.ndarray) and val.shape == array.shape:
-        array = val
-    elif isinstance(val, (tuple, list, np.ndarray)):
-        if len(val) != dim:
-            msg = (f'Wrong dimension {len(val)} provided for value;'
-                   f' expected dimension is {dim}')
-            raise ValueError(msg)
-        if (isinstance(val, np.ndarray) and val.dtype == 'complex128'
-                or np.iscomplex(val).any()):
-            array = np.empty((*mesh.n, dim), dtype='complex128')
-        array[..., :] = val
-    elif callable(val):
-        for index, point in zip(mesh.indices, mesh):
-            res = val(point)
-            try:
-                array[index] = res
-            except TypeError:
-                array = np.array(array, dtype='complex128')
-                array[index] = res
-    elif isinstance(val, dict) and mesh.subregions:
-        if np.iscomplex(list(val.values())).any():
-            array = np.empty((*mesh.n, dim), dtype='complex128')
-        for index, point in zip(mesh.indices, mesh):
-            for region in mesh.subregions.keys():
-                if point in mesh.subregions[region]:
-                    try:
-                        if callable(val[region]):
-                            array[index] = val[region](point)
-                        else:
-                            array[index] = val[region]
-                    except KeyError:
-                        try:
-                            if callable(val['default']):
-                                array[index] = val['default'](point)
-                            else:
-                                array[index] = val['default']
-                        except KeyError:
-                            msg = f"'{region}' or 'default' in value. " \
-                                  + "Either specify all subregions or " \
-                                  + "use key 'default'."
-                            raise KeyError(msg)
-                    break
-            else:
+@functools.singledispatch
+def as_array(val, mesh, dim, dtype):
+    raise ValueError('Unsupported type {type(val)}.')
+
+
+@as_array.register(numbers.Complex)
+@as_array.register(collections.abc.Iterable)
+def _(val, mesh, dim, dtype):
+    # we should think about allowing this as well
+    if isinstance(val, numbers.Complex) and dim > 1 and val != 0:
+        raise ValueError('Wrong dimension 1 provided for value;'
+                         f' expected dimension is {dim}')
```

Comment:
[Repeated from above] We should think about allowing scalar values for vector fields.

Reply:
I can see why allowing this may be confusing. (Somebody may initialise a field with '3e10' - thanking this will act in the x-direction, because they compute everything with respect to an applied field in x. If the actual field value is [3e10, 3e10, 3e10], this is probably not what the user would expect. )